daft/io/_generator.py

```diff
@@ -0,0 +1,99 @@
+# isort: dont-add-import: from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any, Callable, Iterator, List
+
+from daft.daft import Pushdowns, PyTable, ScanOperatorHandle, ScanTask
+from daft.dataframe import DataFrame
+from daft.io.scan import PartitionField, ScanOperator
+from daft.logical.builder import LogicalPlanBuilder
+from daft.logical.schema import Schema
+
+if TYPE_CHECKING:
+    from daft.table.table import Table
+
+
+def _generator_factory_function(
+    func: Callable[[int, Any], Iterator["Table"]], i: int, *args: Any
+) -> Iterator["PyTable"]:
+    for table in func(i, *args):
+        yield table._table
+
+
+def read_generator(
+    generator: Callable[[int, Any], Iterator["Table"]],
+    num_partitions: int,
+    schema: Schema,
+    *generator_args: Any,
```

Comment:
Nit: I feel like this could be simplified to have the caller do its own currying of the callable: `Callable[[int], Iterator["Table"]]` and then we don't need to take in generator_args

Any existing usage of this API should be able to pretty easily use `functools.partial` to achieve the same result

Reply:
Mmmm, very neat!